velox/functions/sparksql/tests/StringTest.cpp

```diff
@@ -497,5 +505,73 @@ TEST_F(StringTest, left) {
   EXPECT_EQ(left("da\u6570\u636Eta", 30), "da\u6570\u636Eta");
 }
 
+TEST_F(StringTest, translate) {
+  EXPECT_EQ(translate("ab[cd]", "[]", "##"), "ab#cd#");
+  EXPECT_EQ(translate("ab[cd]", "[]", "#"), "ab#cd");
+  EXPECT_EQ(translate("ab[cd]", "[]", "#@$"), "ab#cd@");
+  EXPECT_EQ(translate("ab[cd]", "[]", "  "), "ab cd ");
+  EXPECT_EQ(translate("ab\u2028", "\u2028", "\u2029"), "ab\u2029");
+  EXPECT_EQ(translate("abcabc", "a", "\u2029"), "\u2029bc\u2029bc");
+  EXPECT_EQ(translate("abc", "", ""), "abc");
+  EXPECT_EQ(translate("translate", "rnlt", "123"), "1a2s3ae");
+  EXPECT_EQ(translate("translate", "rnlt", ""), "asae");
+  EXPECT_EQ(translate("abcd", "aba", "123"), "12cd");
+  // Test null input.
+  EXPECT_EQ(translate("abc", std::nullopt, "\u2029"), std::nullopt);
+  EXPECT_EQ(translate("abc", "\u2028", std::nullopt), std::nullopt);
+  EXPECT_EQ(translate(std::nullopt, "\u2028", "\u2029"), std::nullopt);
+}
+
+TEST_F(StringTest, translateConstantMatch) {
+  auto rowType = ROW({{"c0", VARCHAR()}});
+  auto exprSet = compileExpression("translate(c0, 'ab', '12')", rowType);
+
+  auto testTranslate = [&](const auto& input, const auto& expected) {
+    auto result = evaluate(*exprSet, makeRowVector({input}));
+    velox::test::assertEqualVectors(expected, result);
+  };
+
+  // Uses ascii batch as the initial input.
+  auto input = makeFlatVector<std::string>({"abcd", "cdab"});
+  auto expected = makeFlatVector<std::string>({"12cd", "cd12"});
+  testTranslate(input, expected);
+
+  // Uses unicode batch as the next input.
+  input = makeFlatVector<std::string>({"abåæçè", "åæçèab"});
+  expected = makeFlatVector<std::string>({"12åæçè", "åæçè12"});
+  testTranslate(input, expected);
+}
+
+TEST_F(StringTest, translateNonconstantMatch) {
+  auto rowType = ROW({{"c0", VARCHAR()}, {"c1", VARCHAR()}, {"c2", VARCHAR()}});
+  auto exprSet = compileExpression("translate(c0, c1, c2)", rowType);
+
+  auto testTranslate = [&](const std::vector<VectorPtr>& inputs,
+                           const auto& expected) {
+    auto result = evaluate(*exprSet, makeRowVector(inputs));
+    velox::test::assertEqualVectors(expected, result);
+  };
+
+  // All inputs are ascii encoded.
+  auto input1 = makeFlatVector<std::string>({"abcd", "cdab"});
```

Comment:
variable names a1, a2, a3 are an anti-pattern; let's rename to input, match and replace